lib/Runtime/Library/JSON.cpp

```diff
@@ -682,56 +681,90 @@ namespace JSON
                 Js::JavascriptStaticEnumerator enumerator;
                 if (object->GetEnumerator(&enumerator, EnumeratorFlags::SnapShotSemantics | EnumeratorFlags::EphemeralReference, scriptContext))
                 {
-                    Js::DynamicObject * dynamicObject = ((Js::DynamicObject*)object);
-                    if (ReplacerFunction != replacerType || !dynamicObject->HasObjectArray())
+                    bool isDynamicObject = Js::DynamicObject::Is(object);
+                    Js::DynamicObject * dynamicObject = nullptr;
+                    bool hasObjectArray = false;
+                    if (isDynamicObject)
                     {
-                        uint32 propertyCount = dynamicObject->GetPropertyCount();
+                        dynamicObject = Js::DynamicObject::FromVar(object);
+                        if (dynamicObject->HasObjectArray())
+                        {
+                            hasObjectArray = dynamicObject->HasObjectArray();
+                        }
+                    }
 
-                        result = Js::ConcatStringBuilder::New(this->scriptContext, propertyCount);
-                        Js::DynamicTypeHandler * typeHandler = dynamicObject->GetTypeHandler();
+                    if (ReplacerFunction != replacerType || !hasObjectArray)
+                    {
+                        uint32 propertyCount;
 
-                        // if object has an objectarray, (aka indexed properties)
+                        // if object has an objectArray, (aka indexed properties)
                         // we need to loop them first since they won't show up on the second loop below
-                        if (dynamicObject->HasObjectArray())
+                        // besides, we might be talking to a non dynamicObject
+                        // in that case, we will also take the enumerator path
+                        if (!isDynamicObject || hasObjectArray)
                         {
+                            // get the precise count because resizing ConcatString will be expensive!
+                            propertyCount = (uint32) this->GetPropertyCount(object, &enumerator);
```

Comment:
Why do you cast to `uint32` here? Doesn't the function already return a `uint32`?